rustls-mbedcrypto-provider/src/kx.rs

```diff
@@ -140,6 +169,101 @@ impl crypto::ActiveKeyExchange for KeyExchange {
     }
 }
 
+struct DheKxGroup {
+    name: NamedGroup,
+    group: ffdhe_groups::FfdheGroup<'static>,
+    priv_key_len: usize,
+}
+
+impl fmt::Debug for DheKxGroup {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:?}", self.name)
+    }
+}
+
+impl SupportedKxGroup for DheKxGroup {
+    fn start(&self) -> Result<Box<dyn ActiveKeyExchange>, Error> {
+        let g = mbedtls::bignum::Mpi::from_binary(self.group.g).map_err(mbedtls_err_to_rustls_error)?;
+        let p = mbedtls::bignum::Mpi::from_binary(self.group.p).map_err(mbedtls_err_to_rustls_error)?;
+
+        let mut rng = super::rng::rng_new().ok_or(rustls::crypto::GetRandomFailed)?;
+        let mut x = vec![0; self.priv_key_len];
+        rng.random(&mut x)
+            .map_err(|_| rustls::crypto::GetRandomFailed)?;
+        let x = mbedtls::bignum::Mpi::from_binary(&x)
+            .map_err(|e| Error::General(format!("failed to make Bignum from random bytes: {}", e)))?;
+
+        let x_pub = g
+            .mod_exp(&x, &p)
+            .map_err(mbedtls_err_to_rustls_error)?;
+
+        Ok(Box::new(DheActiveKeyExchange {
+            name: self.name,
+            group: self.group,
+            x: x.to_binary()
+                .map_err(mbedtls_err_to_rustls_error)?,
+            x_pub: x_pub
+                .to_binary()
+                .map_err(mbedtls_err_to_rustls_error)?,
+        }))
+    }
+
+    fn name(&self) -> NamedGroup {
+        self.name
+    }
+}
+
+struct DheActiveKeyExchange {
+    name: NamedGroup,
+    group: ffdhe_groups::FfdheGroup<'static>,
+    x: Vec<u8>,
```

Comment:
I think use `Mutex<Mpi>` would be faster than converting to/from `vec<u8>`.
I suggest using `Mutex<Mpi>` for `x` and `x_pub`, and `x_pub: Vec<u8>` can remain to speed up the `fn pub_key`

There is a closed PR on `rust-mbedtls` side for add `Sync` impl for `Mpi`, https://github.com/fortanix/rust-mbedtls/pull/197 , after we reopen and merge it, we could get rid of `Mutex` here.

Reply:
Good suggestion. Will do!